L1Trigger/L1THGCal/src/veryfrontend/HGCalVFECompressionImpl.cc

```diff
@@ -1,5 +1,87 @@
 #include "L1Trigger/L1THGCal/interface/veryfrontend/HGCalVFECompressionImpl.h"
 
 HGCalVFECompressionImpl::
-HGCalVFECompressionImpl(const edm::ParameterSet& conf)
-{}
+HGCalVFECompressionImpl(const edm::ParameterSet& conf):
+    exponentBits_(conf.getParameter<uint32_t>("exponentBits")),
+    mantissaBits_(conf.getParameter<uint32_t>("mantissaBits")),
+    rounding_(conf.getParameter<bool>("rounding"))
+{
+  saturable_ = false;
+  saturationValue_ = 0xffffffff;
+  if (((1 << exponentBits_) + mantissaBits_ - 1) < 32) {
+    saturable_ = true;
+    saturationValue_ = (1 << ((1 << exponentBits_) + mantissaBits_ - 1)) - 1;
+  }
+  // TODO: for non-saturating need to get maximum code point as well?
+  // i.e. 8-bit compressed code that would hit 0xffffffff
+
+  // build the compressed values lookup table
+  uint32_t compval = 0;
+  const uint32_t mantissaMask = (1 << mantissaBits_) - 1;
+  uint32_t incval = 1;
+  for (int codeInt = 0; codeInt < 0x100; codeInt++) {
+    const uint8_t code = static_cast<uint8_t>(codeInt);
+    compressedValueLUT_[code] = compval;
+    if ((code & mantissaMask) == 0) {
+      incval <<= 1;
+      if (((code & ~mantissaMask) >> mantissaBits_) <= 1) {
+        incval >>= 1;
+      }
+    }
+    compval += incval;
+  }
+}
+
+uint8_t
+HGCalVFECompressionImpl::
+compressSingle(const uint32_t value)
+{
+  // check for saturation
+  if (saturable_ && value > saturationValue_)
+      return 0xff;
+
+  // count bit length
+  uint32_t valcopy = value;
+  uint8_t bitlen;
+  for (bitlen = 0; valcopy != 0; valcopy >>= 1, bitlen++) {}
+  if (bitlen <= mantissaBits_)
+    return static_cast<uint8_t>(value);
+
+  // build exponent and mantissa
+  const uint8_t exponent = bitlen - mantissaBits_;
+  const uint8_t mantissa = (value >> (exponent-1)) & ~(1<<mantissaBits_);
+
+  // assemble floating-point
+  const uint8_t floatval = (exponent << mantissaBits_) | mantissa;
+
+  // we will never want to round up 0xff here
+  if (!rounding_ || floatval == 0xff) {
+    return floatval;
+  }
+  else {
+    const bool roundup = ((value >> (exponent-2)) & 1) == 1;
+    return roundup ? floatval+1 : floatval;
+  }
+}
+
+uint32_t
+HGCalVFECompressionImpl::
+decompressSingle(const uint8_t code)
+{
+  return compressedValueLUT_[code];
+}
+
+void
+HGCalVFECompressionImpl::
+compress(const std::map<HGCalDetId, uint32_t>& payload,
+               std::map<HGCalDetId, std::array<uint32_t, 2> >& compressed_payload)
+{
+  for (const auto& item : payload) {
+    const uint32_t value = item.second;
+    const uint8_t code = compressSingle(value);
+    const uint32_t compressed_value = decompressSingle(code);
```

Comment:
It should be possible to create the inflated value inside the `compressSingle()` function, on the fly, with shifts and `|` based on the exponent and mantissa values. 
Using this lookup table is a bit confusing because it is not straightforward to understand what is inside.

Reply:
I used lookup table because I am concerned with the code performance should millions of events from a run need to be processed. Otherwise I can change the code to calculate the values instead.

Reply:
Code performance is most likely not an issue here. Other steps in the HGCAL trigger should be much more CPU consuming.